base/essentials.jl

```diff
@@ -169,9 +169,20 @@ end
     for out-of-bounds indices. The user is responsible for checking it manually.
 """
 macro inbounds(blk)
-    :($(Expr(:inbounds,true));
-      $(esc(blk));
-      $(Expr(:inbounds,:pop)))
+    if blk isa Expr && (blk.head === :block || blk.head === :if || blk.head === :let)
+        # Making @inbounds expression-like can incur a performance hit in these
+        # more complicated cases as it assigns a potentially type-unstable value
+        # to an unused variable that isn't currently optimized away. Ref:
+        # https://github.com/JuliaLang/julia/pull/15558#issuecomment-265533514
+        :($(Expr(:inbounds,true));
+          $(esc(blk));
+          $(Expr(:inbounds,:pop)))
+      else
```

Comment:
Hah, looks like my editor was thrown off by the wrapped block spacing.  Good catch, thanks.